internal/rest/resources/resources.go

```diff
@@ -44,68 +45,97 @@ var InternalEndpoints = rest.Resources{
+func ValidateEndpoints(extensionServers []rest.Server, coreAddress string) error {
 	allExistingEndpoints := []rest.Resources{UnixEndpoints, PublicEndpoints, InternalEndpoints}
 	existingEndpointPaths := make(map[string]bool)
+	serverAddresses := map[string]bool{coreAddress: true}
 
+	// Record the paths for all internal endpoints.
 	for _, endpoints := range allExistingEndpoints {
 		for _, e := range endpoints.Endpoints {
 			url := filepath.Join(string(endpoints.PathPrefix), e.Path)
 			existingEndpointPaths[url] = true
 		}
 	}
 
-	for _, e := range extensionServerEndpoints.Endpoints {
-		url := filepath.Join(string(extensionServerEndpoints.PathPrefix), e.Path)
-		if existingEndpointPaths[url] {
-			return fmt.Errorf("Endpoint %q conflicts with internal endpoint", url)
+	for _, server := range extensionServers {
+		// Ensure all servers have resources.
+		if len(server.Resources) == 0 {
+			return fmt.Errorf("Server must have defined resources")
 		}
-	}
 
-	return nil
-}
+		if server.ServeUnix && !server.CoreAPI {
+			return fmt.Errorf("Cannot serve non-core API resources over the core unix socket")
+		}
 
-// GetAndValidateCoreEndpoints extracts all endpoints from extensionServers that should be allocated to the core listener.
-// It also performs the following validations:
-// 1. Only one core API server is allowed.
-// 2. Server configurations are properly set.
-// 3. Path prefixes for endpoints belonging to a single server are not duplicated
-// 4. Enpoints defined in extensionServers do not conflict with internal endpoints.
-func GetAndValidateCoreEndpoints(extensionServers []rest.Server) ([]rest.Resources, error) {
-	var coreEndpoints []rest.Resources
-	var numCoreAPIs int
-
-	for _, extensionServer := range extensionServers {
-		if !extensionServer.CoreAPI {
-			continue
+		if server.CoreAPI && server.Certificate != nil {
+			return fmt.Errorf("Core API server cannot have a pre-defined certificate")
 		}
 
-		numCoreAPIs++
-		if numCoreAPIs > 1 {
-			return nil, fmt.Errorf("Only one core API server is allowed")
+		if server.CoreAPI && server.Address != (types.AddrPort{}) {
+			return fmt.Errorf("Core API server cannot have a pre-defined address")
+		}
+
+		// Ensure all servers with a defined address are unique.
+		if server.Address != (types.AddrPort{}) {
+			if serverAddresses[server.Address.String()] {
+				return fmt.Errorf("Server address %q conflicts with another Server", server.Address.String())
+			}
+
+			serverAddresses[server.Address.String()] = true
+		} else if server.Protocol != "" {
+			return fmt.Errorf("Server protocol defined without address")
 		}
 
-		err := extensionServer.ValidateServerConfigs()
+		// Ensure no endpoint path conflicts with another endpoint on the same server.
+		// If a server lacks an address, we need to compare it to every other server
+		// that also lacks an address, as well as the internal endpoints.
+		serverPaths, err := resourcesConflict(server, existingEndpointPaths)
```

Comment:
@masnax just a thought here, if we restrict the number of servers with `CoreAPI=true` to 1, then we wouldn't need to do this validation? From a usage perspective I can't really see a reason why multiple server structs would be defined with `CoreAPI=true`. Although I think maybe we also shouldn't assume what the user might want to do, so feel free to disregard this comment :)